Chapters/2-marshalling.pillar

```diff
@@ -1,96 +1,123 @@
+- A C integer value (representable in 31/61 bits or less) transformed to a Pharo ==SmallInteger== needs to be shifted to the left and incremented, ==(2r10 << 1) + 1==, transforming ==2r10== to ==2r101==.
 
-This resulting ''representation mismatch'' requires that the uFFI transform small integers to C ==int==s (and vice\-versa) as follows:
-- a Pharo small integer transformed to a C value needs to be logically shifted to the right, ==2r101 >> 1==, transforming ==2r101== to ==2r10==.
-- a C integer (representable in 31 bits or less) transformed to a Pharo small integer needs to be shifted to the left and incremented, ==(2r10 << 1) + 1==, transforming ==2r10== to ==2r101==.
+Each type\-to\-type transformation has its own particular rule that uFFI follows to ensure that correct representation is always maintained.
 
 !!!! Pharo-to-C Marshalling
 
 When marshalling Pharo objects to C, uFFI decides the transformation rule to use depending on two pieces of information.
 First, it considers the concrete type of the marshalled Pharo object, ''its class''.
 Second, it considers the C type defined in the function binding as the target transformation type.
-At run time, when the binding method is executed, uFFI reads the type of the binding argument and transforms the argument object into the indicated C type representation, performing a type cast/coercion as necessary.
+At run time, when the binding method is executed, uFFI reads the type of the Pharo argument and transforms the argument object into the indicated C type representation, performing a type cast/coercion as necessary.
 
-These transformation rules have several consequences, which we will illustrate with several cases, using our previous ==abs:== binding example.
+These transformation rules can have consequences, which we illustrate with the following cases, using our previous ==abs:== binding example\:
 
-- ==SmallInteger==s are transformed to ==int==s as expected (since they cannot overflow)\:
+- ==SmallInteger==s in 32-bit Pharo images are transformed to ==int==s as expected (since 31 bits cannot overflow into 32 bits)\:
 
 [[[language=smalltalk
 FFITutorial abs: -42.
 => 42
 ]]]
 
-- Pharo integers that overflow the C ==int== size are coerced by truncating to size, producing results similar to what a C program would produce\:
+[[[language=smalltalk
+FFITutorial abs: SmallInteger maxVal negated.
+=> 1073741823
+]]]
+
+- ==SmallInteger==s in 64-bit Pharo images ''can'' overflow the size of a C ==int== (still 32 bits on most 64-bit hosts), and so are coerced by truncating their value to fit, producing results similar to what a C program would produce in a similar situation.  Since the ==maxVal== of a 64-bit Pharo ==SmallInteger== is 60 \'1\' bits (plus a sign bit), it truncates to 32 \'1\' bits, which, to C, is the two\'s complement value -1.  Hence,
 
 [[[language=smalltalk
-FFITutorial abs: SmallInteger maxVal.  "Cast to all 1's"
+FFITutorial abs: SmallInteger maxVal negated.
```

Comment:
Yes, nice example!

Now, this assumes the reader knows that ==int==s are 32bit sized :) Maybe we can add a clarification on this?

Reply:
And also the fact that ==int== is signed, thus 1 bit is there for the sign :)

Reply:
(which is necessary to understand the explanation on the return value in the next subsection ;))

Reply:
Yes, I've been trying to add notes to remind the reader that C int's are 32 bit regardless of the Pharo image bit size.  And I mentioned that "all 1s" is two's complement of "-1". 

One thing we might mention is that Pharo keeps a separate sign bit, too.  I found this out while experimenting: `maxVal` and `maxVal negated` have the _same_ number of value bits, which is not like C 2's complement values.  The reader probably should know this, too.